File: py/2022/day04.py
from aoc import main

Pair = tuple[set[int], set[int]]


def parse(s: str) -> list[Pair]:
    rv: list[Pair] = []
    for line in s.splitlines():
        a, b = line.split(",", 2)
        a_start, a_end = [int(x) for x in a.split("-", 2)]
        b_start, b_end = [int(x) for x in b.split("-", 2)]
        rv.append((set(range(a_start, a_end + 1)), set(range(b_start, b_end + 1))))
    return rv


if __name__ == "__main__":
    main(
        lambda s: len([1 for a, b in parse(s) if a <= b or b <= a]),
        lambda s: len([1 for a, b in parse(s) if a & b]),
    )

File: ts/2022/day04.ts
import { main } from 'lib/advent';
import { Iter, iter } from 'lib/iter';
import { lines, range } from 'lib/util';

type Range = Set<number>;

function parse(lines: string[]): Iter<Range[]> {
  return iter(lines).map((line) =>
    line.split(',', 2).map((r) => {
      const [a, b] = r.split('-', 2).map(Number);
      return new Set(range(a, b + 1));
    })
  );
}

function contains<T>(a: Set<T>, b: Set<T>): boolean {
  return [...a].every((x) => b.has(x)) || [...b].every((x) => a.has(x));
}

function overlaps<T>(a: Set<T>, b: Set<T>): boolean {
  return [...a].some((x) => b.has(x));
}

main(
  (s) => parse(lines(s)).filter(([a, b]) => contains(a, b)).size,
  (s) => parse(lines(s)).filter(([a, b]) => overlaps(a, b)).size
);
